doc/source/user/how-to-io.rst

```diff
@@ -0,0 +1,323 @@
+        ("format", "S4"),         # 4-byte string
+        ("fmt_id", "S4"),
+        ("fmt_size", "<u4"),
+        ("audio_fmt", "<u2"),     #
+        ("num_channels", "<u2"),  # .. more of the same ...
+        ("sample_rate", "<u4"),   #
+        ("byte_rate", "<u4"),
+        ("block_align", "<u2"),
+        ("bits_per_sample", "<u2"),
+        ("data_id", "S4"),
+        ("data_size", "u4"),
+        #
+        # the sound data itself cannot be represented here:
+        # it does not have a fixed size
+       ])
+
+    data = np.fromfile(f,dtype=wave_header_dtype)
+
+Credit: Pauli Virtanen, :ref:`advanced_numpy`.
+
+.. _how-to-io-large-arrays:
+
+******************************************************************************
+Write or read large arrays
+******************************************************************************
+
+**Arrays too large to fit in memory** can be treated like ordinary in-memory
+arrays using :func:`numpy.memmap`::
+
+    import numpy as np
+    array = np.memmap("mydata/myarray.arr", mode="r", dtype=np.int16, shape=(1024, 1024))
+
+Memory-mapping functionality is also available with `numpy.load` via the
+``mmap_mode`` keyword argument.
+
+::
+
+  large_array[some_slice] = np.load('path/to/small_array', mmap_mode='r')
+
+Memory-mapped can be used for efficient disk I/O, but lacks features like data
+chunking and compression. For finer control over data access, consider more
+full-featured file formats or data libraries such as ``HDF5`` or ``Zarr``.
+
+* **HDF5**: `h5py <https://www.h5py.org/>`_ or `PyTables <https://www.pytables.org/>`_.
+* **Zarr**: `here <https://zarr.readthedocs.io/en/stable/tutorial.html#reading-and-writing-data>`_.
+* **NetCDF**: :class:`scipy.io.netcdf_file`.
+
+For details on tradeoffs among memmap, Zarr, and HDF5, see
+`pythonspeed.com <https://pythonspeed.com/articles/mmap-vs-zarr-hdf5/>`_.
+
+******************************************************************************
+Write files for reading by other (non-NumPy) tools
+******************************************************************************
+
+Formats for **exchanging data** with other tools include HDF5, Zarr, and
+NetCDF (see :ref:`how-to-io-large-arrays`).
+
+******************************************************************************
+Write or read a JSON file
+******************************************************************************
+
+NumPy arrays are **not** directly
+`JSON serializable <https://github.com/numpy/numpy/issues/12481>`_.
+
+
+.. _how-to-io-pickle-file:
+
+******************************************************************************
+Save/restore using a pickle file
+******************************************************************************
+
+Not recommended, due to lack of security and portability.
+
+ * **security**: not secure against erroneous or maliciously constructed data
+ * **portability**: may not be loadable on different Python installations

```

Comment:
Numpy tries quite hard to make this not the case, and `np.save` may not be any better about this. I don't think this is really a concern.

Reply:
> make this not the case

Which are you saying is not the case -- security, portability, or both?

Reply:
Portability is what I'm referring to in this comment.